### PR TITLE
Add lint test for merge markers

### PR DESCRIPTION
**Pull request by KevinMenden**
## PR checklist

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` is updated
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [x] Documentation in `docs` is updated

Adds a lint test looking for merge markers.

Closes https://github.com/nf-core/tools/issues/321

### CODE REVIEW

**Comment by KevinMenden** [comment]
Accidentally made this PR against master (again 🤦 ) that's why the branch protection failed....

**Comment by ewels** [comment]
Hmm strange - the comment still seemed to reference the wrong repo name, which should have been fixed in https://github.com/nf-core/tools/pull/860

**Comment by ewels** [comment]
Ah wait, the PR is against `master` so I guess that’s where the test is from as well so will need to wait for the release for the fix to be active..

**Comment by KevinMenden** [comment]
Okay good was already wondering why the tests still fail after running again 🤔

File: nf_core/lint/merge_markers.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            try:
+                with io.open(os.path.join(root, fname), "rt", encoding="latin1") as fh:
+                    for l in fh:
+                        if ">>>>>>>" in l:
+                            warned.append(f"Merge marker in `{fname}`: {l}")
+                        if "<<<<<<<" in l:
+                            warned.append(f"Merge marker in `{fname}`: {l}")
+            except FileNotFoundError:
+                log.debug(f"Could not open file {fname} in merge_markers lint test")
+    return {"passed": passed, "warned": warned, "failed": failed}
```
Comment (ewels) [inline nf_core/lint/merge_markers.py:43]:
```suggestion
    if len(failed) == 0:
        passed.append("No merge markers found in pipeline files")
    return {"passed": passed, "failed": failed}
```

File: nf_core/lint/merge_markers.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+def merge_markers(self):
+    """Check for remaining merge markers.
+
+    This test looks for remaining merge markers in the code, e.g.:
+    >>>>>>> or <<<<<<<
+
+
+    """
+    passed = []
+    warned = []
```
Comment (ewels) [inline nf_core/lint/merge_markers.py:20]:
```suggestion
```

File: nf_core/lint/merge_markers.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        # Ignore files
+        for i in ignore:
+            dirs = [d for d in dirs if not fnmatch.fnmatch(os.path.join(root, d), i)]
+            files = [f for f in files if not fnmatch.fnmatch(os.path.join(root, f), i)]
+        for fname in files:
+            try:
+                with io.open(os.path.join(root, fname), "rt", encoding="latin1") as fh:
+                    for l in fh:
+                        if ">>>>>>>" in l:
+                            warned.append(f"Merge marker in `{fname}`: {l}")
```
Comment (ewels) [inline nf_core/lint/merge_markers.py:38]:
```suggestion
                            failed.append(f"Merge marker in `{fname}`: {l}")
```

File: nf_core/lint/merge_markers.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                        if "<<<<<<<" in l:
+                            warned.append(f"Merge marker in `{fname}`: {l}")
```
Comment (ewels) [inline nf_core/lint/merge_markers.py:40]:
```suggestion
                            failed.append(f"Merge marker in `{fname}`: {l}")
```

File: tests/lint/merge_markers.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        main_nf_content = fh.read()
+    main_nf_content = ">>>>>>>\n" + main_nf_content
+    with open(os.path.join(new_pipeline, "main.nf"), "w") as fh:
+        fh.write(main_nf_content)
+
+    lint_obj = nf_core.lint.PipelineLint(new_pipeline)
+    lint_obj._load()
+
+    results = lint_obj.merge_markers()
+    assert len(results["warned"]) > 0
```
Comment (ewels) [inline tests/lint/merge_markers.py:22]:
```suggestion
    assert len(results["failed"]) > 0
    assert len(results["passed"]) == 0
```

File: tests/lint/merge_markers.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    assert "Merge marker in `main.nf`: >>>>>>>\n" == results["warned"][0]
```
Comment (ewels) [inline tests/lint/merge_markers.py:23]:
```suggestion
    assert "Merge marker in `main.nf`: >>>>>>>\n" == results["failed"][0]
```

**Review by ewels — APPROVED** [review]
I think this should trigger a failure, not just a warning (so that it blocks PR merges).

Also suggested adding a pass if there are no problems found.

**Comment by BarryDigby** [comment]
Hi guys, just letting you know that the test fails on `.png` images, see below: 

```console
 merge_markers: Merge marker '<<<<<<<' in ./docs/images/output/DESeq2_QC/circRNA/DESeq2_dispersion.png: 4¨W¯^Ú\Oqqñ#G.]ºSXXhiiùl³<<<<<<<==   #GÜ¸q#))éÑ£GÅÅÅ¦¦¦666666¯¿þºWïÞ½ÍÍÍ[YtA… │
```

```console
grep -R "<<<<<<<" docs/images/
Binary file docs/images/output/DESeq2_QC/circRNA/DESeq2_dispersion.png matches
```

**Comment by KevinMenden** [comment]
Thanks! Didn't see that one coming ... @ewels maybe omit anything that ends with `.png` , `.jpg` etc?
Might be generally useful to have a list of file endings that are ignored.

**Pull request merged**